reference/data-consistency.md

```diff
@@ -0,0 +1,78 @@
+---
+description: Description of Turso data consistency guarantees.
+keywords:
+  - turso
+  - technical
+  - reference
+  - consistency
+---
+
+# Data consistency
+
+Turso is built on top of [libSQL], which is a fork of [SQLite]. SQLite, as an
+embedded database, offers a strictly serializable consistency model. However,
+libSQL, as a network-accessible and replicated database, can not offer such
+a strong guarantee.
+
+## Connections
+
+Clients may establish a connection to a sqld instance, and each instance may
+handle a number of client connections concurrently. For the purpose of this
+documentation, a connection to a sqld instance is referred to as a *process*,
+which issues an ordered sequence of queries with results.
+
+## Read and write behavior
+
+### On the primary instance
+
+All operations by processes connected to the [primary] are linearizable, forming
+a ordered sequence of completed operations (a history).
+
+### On the replicas
+
+#### Writes are forwarded to the primary
+
+All writes from a process connected to a [replica] are automatically forwarded
+to the primary and applied there. Changes to the primary are eventually pulled
+from each replica and made available locally. A process is guaranteed to see all
+writes that happened on the primary up until (at least) the last write performed
+by the process. There are no hard guarantees on how long it will take for a
+replica to observe a change from the primary, considering network latency and
+availability, and total load on the primary.
+
+#### Reads are local and not globally ordered
+
+All reads by a process come from the locally replicated database. Peer replicas
+are not necessarily all in sync with each other. Reads from different replicas
+can yield different data until each replica receives the latest changes from the
+primary. Furthermore, concurrent reads on a replica may observe snapshots of
+data from different points in time.
+
+### Monotonic reads
+
+All database instances offer monotonic reads. A read operation will never appear
+to yield data older than a prior read on the same instance.
+
+## Transactional consistency
+
+A sqld transaction ([batch] or [interactive transaction]) is equivalent to a
+[SQLite transaction] and observes its semantics. Each process may have one
+transaction ongoing.
```

Comment:
Nothing prevents the client from opening multiple transactions on the same Hrana connection. Read-only batches are executed concurrently, so many of them can be "ongoing". However, write batches and interactive transactions are serialized on the primary, so there can be only one ongoing transaction *globally*, not just one per client.